src/coreclr/vm/interpexec.cpp

```diff
@@ -1186,15 +1186,46 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
                     callArgsOffset = ip[2];
                     methodSlot = ip[3];
 
-                    OBJECTREF objRef = AllocateObject((MethodTable*)pMethod->pDataItems[ip[4]]);
+                    MethodTable *pClass = (MethodTable*)pMethod->pDataItems[ip[4]];
+                    // FIXME: Duplicated code from CALL_INTERP_SLOT
+                    size_t targetMethod = (size_t)pMethod->pDataItems[methodSlot];
+                    MethodDesc *pMD = nullptr;
+                    if (targetMethod & INTERP_METHOD_HANDLE_TAG)
+                    {
+                        pMD = (MethodDesc*)(targetMethod & ~INTERP_METHOD_HANDLE_TAG);
+                    }
 
-                    // This is return value
-                    LOCAL_VAR(returnOffset, OBJECTREF) = objRef;
-                    // Set `this` arg for ctor call
-                    LOCAL_VAR (callArgsOffset, OBJECTREF) = objRef;
-                    ip += 5;
+                    // If we are constructing a type with a component size (i.e. a string) its constructor is a special
+                    //  fcall that is basically a static method that returns the new instance.
+                    if (pMD && pClass->HasComponentSize())
+                    {
+                        // The compiler didn't know about this so it reserved space for a this-reference. We need to skip
```

Comment:
This sounds like a temporary workaround. The compiler can know about this (by checking `CORINFO_FLG_VAROBJSIZE` flag). What needs to happen to move this logic to the compiler?

Reply:
m_compHnd didn't appear to expose the things I needed to determine this. I can take another look.

Reply:
That's the flag `CORINFO_FLG_VAROBJSIZE` that @jkotas mentioned offline yesterday:
https://github.com/dotnet/runtime/blob/fc85a87a0c41f6e7940f7f24a324408aafdb86d2/src/coreclr/jit/importercalls.cpp#L1010-L1026
You can get it by `getClassAttribs` or by the `getCallInfo` in the `CORINFO_CALL_INFO::classFlags`.

Reply:
Updated. https://github.com/dotnet/runtime/pull/115914/commits/0298d2e5b9448d68bb55939df08a79f0d49dd65d
Not sure how I feel about it.